source/slang/slang-ir-spirv-legalize.cpp

```diff
@@ -0,0 +1,258 @@
+            if (oldResultType->getAddressSpace() != ptrType->getAddressSpace())
+            {
+                IRBuilder builder;
+                builder.sharedBuilder = &m_sharedContext->m_sharedIRBuilder;
+                builder.setInsertBefore(inst);
+                auto newPtrType = builder.getPtrType(
+                    oldResultType->getOp(),
+                    oldResultType->getValueType(),
+                    ptrType->getAddressSpace());
+                auto newInst =
+                    builder.emitFieldAddress(newPtrType, inst->getBase(), inst->getField());
+                inst->replaceUsesWith(newInst);
+                inst->removeAndDeallocate();
+                addUsersToWorkList(newInst);
+            }
+        }
+    }
+
+    void processStructuredBufferType(IRHLSLStructuredBufferTypeBase* inst)
+    {
+        IRBuilder builder;
+        builder.sharedBuilder = &m_sharedContext->m_sharedIRBuilder;
+        builder.setInsertBefore(inst);
+        auto arrayType = builder.getUnsizedArrayType(inst->getElementType());
+        auto ptrType = builder.getPtrType(kIROp_PtrType, arrayType, SpvStorageClassStorageBuffer);
+        inst->replaceUsesWith(ptrType);
+        inst->removeAndDeallocate();
+        addUsersToWorkList(ptrType);
+    }
+
+    void propagate()
+    {
+        addToWorkList(m_module->getModuleInst());
+        while (workList.Count() != 0)
+        {
+            IRInst* inst = workList.getLast();
+            workList.removeLast();
+            switch (inst->getOp())
+            {
+            case kIROp_GlobalParam:
+                processGlobalParam(as<IRGlobalParam>(inst));
+                break;
+            case kIROp_GlobalVar:
+                processGlobalVar(as<IRGlobalVar>(inst));
+                break;
+            case kIROp_Call:
+                processCall(as<IRCall>(inst));
+                break;
+            case kIROp_getElementPtr:
+                processGetElementPtr(as<IRGetElementPtr>(inst));
+                break;
+            case kIROp_FieldAddress:
+                processFieldAddress(as<IRFieldAddress>(inst));
+                break;
+            case kIROp_HLSLStructuredBufferType:
+            case kIROp_HLSLRWStructuredBufferType:
+                processStructuredBufferType(as<IRHLSLStructuredBufferTypeBase>(inst));
+                break;
+            default:
+                for (auto child = inst->getLastChild(); child; child = child->getPrevInst())
+                {
+                    addToWorkList(child);
+                }
+                break;
+            }
+        }
+    }
+};
+
+void propagateStorageClass(SPIRVEmitSharedContext* sharedContext, IRModule* module)
+{
+    StorageClassPropagationContext context(sharedContext, module);
+    context.propagate();
+}
+
+void legalizeIRForSPIRV(
+    SPIRVEmitSharedContext* context,
+    IRModule* module,
+    DiagnosticSink* sink)
+{
+    SLANG_UNUSED(sink);
+    propagateStorageClass(context, module);
+}
```

Comment:
Would it be possible to make this pass not be explicitly SPIR-V related? Could we define a Slang-side `enum` for storage classes and emit a propagation pass like this for them without any reference to SPIR-V or its rules, and then once that it done just translate the Slang storage classes to the SPIR-V ones as part of emission?

I guess the question I'm asking here is whether there are multiple targets that might need/want storage class propagation and, if so, whether they can all share a common propagation pass and `enum` for the storage classes?

Realistically, it is probably best to leave this logic exactly as it is here until we actually have a second use case and worry about the details then.